cpp/src/arrow/util/reflection_internal.h

```diff
@@ -91,21 +92,260 @@ PropertyTuple<Properties...> MakeProperties(Properties... props) {
+  [[nodiscard]] constexpr const T& operator[](size_t i) const { return begin()[i]; }
+
+ private:
+  T values_[N] = {};  // NOLINT
 };
 
-template <typename T, typename Enable = void>
-struct has_enum_traits : std::false_type {};
+template <typename T, size_t N, size_t M>
+constexpr bool operator==(const array<T, N>& l, const array<T, M>& r) {
+  if constexpr (N != M) {
+    return false;
+  } else {
+    for (size_t i = 0; i < N; ++i) {
+      if (l[i] != r[i]) return false;
+    }
+    return true;
+  }
+}
+
+template <typename T, size_t N, typename R>
+constexpr bool operator!=(const array<T, N>& l, const R& r) {
+  return !(l == r);
+}
+
+template <auto...>
+struct sequence {};
+
+constexpr std::string_view TrimNamespace(std::string_view name) {
+  for (size_t i = name.size(); i != 0; --i) {
+    char c = name[i - 1];
+    if ('A' <= c && c <= 'Z') continue;
+    if ('a' <= c && c <= 'z') continue;
+    if ('0' <= c && c <= '9') continue;
+    if (c == '_') continue;
+    name = name.substr(i);
+    break;
+  }
+  return name;
+}
+
+template <auto Value>
+constexpr auto kValueNameStorage = [] {
+  constexpr std::string_view name = [] {
+    std::string_view name = pretty_function<Value>();
+    auto [prefix, suffix] = kValueNamePrefixSuffix;
+    name.remove_prefix(prefix);
+    name.remove_suffix(suffix);
+    name = TrimNamespace(name);
+    return name;
+  }();
+
+  // copying out of the string_view here ensures that characters which were
+  // sliced out are not present in (release)binaries
+  return array<char, name.size()>{name.data()};
+}();
 
 template <typename T>
-struct has_enum_traits<T, void_t<typename EnumTraits<T>::Type>> : std::true_type {};
+constexpr auto kTypeNameStorage = [] {
+  constexpr std::string_view name = [] {
+    std::string_view name = pretty_function<T>();
+    auto [prefix, suffix] = kTypeNamePrefixSuffix;
+    name.remove_prefix(prefix);
+    name.remove_suffix(suffix);
+    return TrimNamespace(name);
+  }();
+  return array<char, name.size()>{name.data()};
+}();
 
-}  // namespace internal
-}  // namespace arrow
+template <typename Enum, int I, Enum E = static_cast<Enum>(I)>
+constexpr bool IsValidEnumMember() {
+  return kValueNameStorage<E>[0] < '0' || kValueNameStorage<E>[0] > '9';
+}
+
+template <typename Enum,
+          typename Limits = std::numeric_limits<std::underlying_type_t<Enum>>,
+          int I = Limits::min(), int Max = Limits::max(), Enum... Members>
+struct DefaultEnumMembers;
+
+template <typename Enum, typename Limits, int Max, Enum... Members>
+struct DefaultEnumMembers<Enum, Limits, Max, Max, Members...> {
+  using members = sequence<Members...>;
+};
+
+template <typename Enum, typename Limits, int I, int Max, Enum... Members>
+struct DefaultEnumMembers
+    : std::conditional_t<
+          IsValidEnumMember<Enum, I>(),
+          DefaultEnumMembers<Enum, Limits, I + 1, Max, Members..., static_cast<Enum>(I)>,
+          DefaultEnumMembers<Enum, Limits, I + 1, Max, Members...>> {};
```

Comment:
Hmm... this seems to lead to O(N) compile-time complexity where N is `std::numeric_limits<std::underlying_type_t<Enum>>::max() - std::numeric_limits<std::underlying_type_t<Enum>>::min()`. That sounds concerning, especially if compile-time evaluation is slow (which it may very well be, depending on the compiler).

Reply:
That's why this is limited to enumerations which are only one byte wide, so we'll do a maximum of 256 iterations per enum. I wouldn't expect that to have a heavy compile time cost